sdl2promela/promela/generator.py

```diff
@@ -0,0 +1,182 @@
+from multipledispatch import dispatch
+from typing import List, Set, Text, Tuple, Type, TextIO
+
+from . import model
+
+INDENT = "  "
+
+class Context:
+    stream : TextIO
+    indents : List[str]
+    pending_indent : bool
+
+    def __init__(self, stream : TextIO):
+        self.stream = stream
+        self.indents = []
+        self.pending_indent = False
+
+    def push_indent(self, indent : str):
+        self.indents.append(indent)
+
+    def pop_indent(self):
+        self.indents.pop()
+
+    def _get_indent(self):
+        return "".join((self.indents))
+
+    def output(self, data : str):
+        if self.pending_indent:
+            print(self._get_indent(), file=self.stream, end='')
+            self.pending_indent = False
+        lines = data.splitlines(keepends=True)
+        line_count = len(lines)
+        for i in range(0, line_count):
+            line = lines[i]
+            print(line, file=self.stream, end='')
+            if line.endswith("\n"):
+                if i == line_count - 1:
+                    self.pending_indent = True
+                else:
+                    print(self._get_indent(), file=self.stream, end='')
+
+class StatementsWrapper:
+    statements : List[model.Statement]
+
+    def __init__(self, statements : List[model.Statement]):
+        self.statements = statements
+
+@dispatch
+def generate(context, element):
+    raise NotImplementedError("generate() not implemented for " + element)
+
+@dispatch(Context, model.BinaryExpression)
+def generate(context : Context, expression : model.BinaryExpression):
+    context.output("(")
+    generate(context, expression.left)
+    if expression.operator == model.BinaryOperator.ADD:
+        context.output(" + ")
+    elif expression.operator == model.BinaryOperator.SUBTRACT:
+        context.output(" - ")
+    elif expression.operator == model.BinaryOperator.MULTIPLY:
+        context.output(" * ")
+    elif expression.operator == model.BinaryOperator.DIVIDE:
+        context.output(" / ")
+    elif expression.operator == model.BinaryOperator.EQUAL:
+        context.output(" == ")
+    elif expression.operator == model.BinaryOperator.LESS:
+        context.output(" < ")
+    elif expression.operator == model.BinaryOperator.GREATER:
+        context.output(" > ")
+    elif expression.operator == model.BinaryOperator.LEQUAL:
+        context.output(" <= ")
+    elif expression.operator == model.BinaryOperator.GEQUAL:
+        context.output(" >= ")
+    generate(context, expression.right)
+    context.output(")")
+
+@dispatch(Context, model.Call)
+def generate(context : Context, call : model.Call):
+    context.output(call.target)
+    context.output("(")
+    last_index = len(call.parameters) -1
+    for i in range(0, len(call.parameters)):
```

Comment:
str join?

of `for param in params[:-1]` and special print of last param

I just rarely see `for (int i = 0; ...` nowadays ;)

Reply:
Except generate does not return an output, just writes to the stream, so there is no list of strings to join....

Reply:
Best I could do is using [:-1], but then I issue the call twice...